fastai/text/data.py

```diff
@@ -13,44 +13,106 @@
+        #         when iterating backwards then ei becomes 1+ the last token to be extraced in the rag 
+        #overlap: overlap=1 between batches, because we only predict the next token        
+        self.idx   = LanguageModelPreLoader.CircularIndex(len(self.dataset.x.items), not self.backwards)
+        self.batch = np.zeros( (self.bs, self.bptt+1), dtype=np.int64)
+        self.x, self.y = self.batch[:,0:self.bptt], self.batch[:,1:self.bptt+1]      
+        self.ei    = np.zeros(self.bs, dtype=np.int)
+        self.eo    = np.zeros(self.bs, dtype=np.int)
+
     def on_epoch_begin(self, **kwargs):
-        self.idxs = np.random.permutation(len(self.dataset)) if self.shuffle else arange_of(self.dataset)
-        self.text_idx = np.concatenate([[0],self.lengths[self.idxs].cumsum()])
-        
+        if self.idx is None: self.allocate_buffers()
+        if self.shuffle:     self.idx.shuffle()
+        self.idx.forward(self.backwards is False) 
+
+        #set up ei and eo to index the data so batches has continuous rows  
+        step   = self.totalToks / self.bs
+        ln_rag = countTokens = 0
+        i_rag  = -1
+        for i in range(0,self.bs):
+            while ln_rag <= int(step * i) - countTokens :
+                countTokens += ln_rag
+                i_rag       += 1
+                ln_rag       = len( self.dataset.x.items[self.idx[i_rag]] )
+            self.ei[i] = i_rag
+            self.eo[i] = ( ln_rag - int(step * i - countTokens) ) if self.backwards else int(step * i - countTokens)
+
     #Training dl gets on_epoch_begin called, val_dl, on_epoch_end
     def on_epoch_end(self, **kwargs): self.on_epoch_begin()
-    
+
     def __getitem__(self, k:int):
         if self.item is not None: return self.dataset[0]
-        if not hasattr(self, 'idxs'): self.on_epoch_begin()
-        #The dataloader will send (batch_index) * bs + sample_index, converting to where to read in the stream 
-        read_idx = (k//self.bs) * self.bptt
-        i = read_idx  + (k % self.bs) * self.n 
-        seq_len = min(self.bptt, self.n-read_idx-1)
-        #Getting the indexes of the texts that start and finish the portion i---i+seq_len in the stream 
-        start,end = np.argmax(self.text_idx >= i)-1,np.argmin(self.text_idx <= i+seq_len+1)
-        start = max(0,start)
-        #Grabbing the texts we need in the dataset
-        if self.backwards: concat = np.concatenate([self.dataset.x.items[j][::-1] for j in self.idxs[start:end]])
-        else: concat = np.concatenate([self.dataset.x.items[j] for j in self.idxs[start:end]])
-        #Returning the right portion.
-        start_idx = i-self.text_idx[start]
-        return concat[start_idx:start_idx+seq_len], concat[start_idx+1:start_idx+seq_len+1]
+        if self.idx is None:      self.on_epoch_begin()
+
+        j = k % self.bs
+        if self.backwards: self.ei[j],self.eo[j] = self.fill_backward(  self.dataset.x.items, self.idx, self.batch[j], 
```

Comment:
`fill_backward` and `fill_forward` only have two lines of code that differ. They should be put together as `fill` with a `backward` argument, that would help refactor a lot of dupe code.